src/main.cairo

```diff
@@ -1,32 +1,231 @@
+}
+
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
+@external
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
+) {
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
+
+    let (prev_renew) = is_renewing.read(caller, domain);
+    is_renewing.write(caller, domain, 1 - prev_renew);
+
+    toggled_renewal.emit(domain, caller, 1 - prev_renew);
+
+    return ();
+} 
+
+// @notice Renew a domain
+// @param domain Domain to renew
+// @dev callable by anyone, but will only renew if renewals are activated for domain
+@external
+func renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    root_domain: felt, renewer: felt,
+) {
+    alloc_locals;
+    let (naming) = naming_contract.read(); 
+    let (can_renew) = is_renewing.read(renewer, root_domain);
+
+    with_attr error_message("Renewer has not activated renewals for this domain") {
+        assert_not_zero(can_renew);
+    }
+
+    let (expiry) = Naming.domain_to_expiry(naming, 1, cast(new(root_domain), felt*));
+    let (block_timestamp) = get_block_timestamp();
+    with_attr error_message("Domain is not set to expire within a month") {
+        assert_le(expiry, block_timestamp + 86400 * 30);
+    }
+
+    let (pricing) = pricing_contract.read();
+    let (erc20, renewal_price) = Pricing.compute_renew_price(pricing, root_domain, 365);
+
+    let (contract) = get_contract_address();
+    with_attr error_message("Error transferring tokens from renewer to contract") {
+        IERC20.transferFrom(erc20, renewer, contract, renewal_price);
+    }
+
+    Naming.renew(naming, root_domain, 365);
+    
+    domain_renewed.emit(root_domain, renewer, 365);
+
+    return ();
+}
+
+@external
+func batch_renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain_len: felt, domain: felt*, renewer_len: felt, renewer: felt*
+) {
+    with_attr error_message("domain and renewer array must have same length") {
+        assert domain_len = renewer_len;
+    }
+    return _batch_renew_iter(domain_len, domain, renewer_len, renewer);
+}
+
+func _batch_renew_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain_len: felt, domain: felt*, renewer_len: felt, renewer: felt*
+) {
+    if (domain_len == 0) {
+        return ();
+    }
+    renew(domain[0], renewer[0]);
+    return _batch_renew_iter(domain_len - 1, domain + 1, renewer_len - 1, renewer + 1);
+}
+
+// @notice Vote for an upgrade
+// @param upgrade Upgrade to vote for
+@external
+func vote_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    upgrade_id: felt, implementation_hash: felt, vote: felt
+) {
+    let (caller) = get_caller_address();
+
+    with_attr error_message("Votes can either be 1 or 0") {
```

Comment:
This is a good way to do it, but I am curious: why not just writing (1-current_value)?

Reply:
Up if not resolved @irisdv